mlir-tensorrt/tensorrt/lib/TensorRT/Transforms/TransposeReshapeElimination.cpp

```diff
@@ -0,0 +1,2568 @@
+      }
+    }
+    if (!hasTransposeInput)
+      return failure();
+
+    SmallVector<Value> newInputs;
+    for (size_t i = 0; i < op.getInputs().size(); i++) {
+      auto input = op.getInputs()[i];
+      if (auto transpose = input.getDefiningOp<tensorrt::TransposeOp>()) {
+        auto perm = transpose.getPermutation();
+        SmallVector<int64_t> equation;
+        for (char c : einsumEquationLhsParts[i]) {
+          equation.push_back(c);
+        }
+
+        equation = inversePermutation(perm).compose(equation);
+        einsumEquationLhsParts[i] = "";
+        for (int j = 0; j < equation.size(); j++) {
+          einsumEquationLhsParts[i] += (char)equation[j];
+        }
+        newInputs.push_back(transpose.getInput());
+      } else {
+        newInputs.push_back(input);
+      }
+    }
+
+    std::string newEinsumEquation = einsumEquationLhsParts[0];
+    for (int i = 1; i < einsumEquationLhsParts.size(); i++) {
+      newEinsumEquation += "," + einsumEquationLhsParts[i];
+    }
+    newEinsumEquation += "->" + einsumEquationRhs;
+
+    rewriter.replaceOpWithNewOp<tensorrt::EinsumOp>(op, op.getType(), newInputs,
+                                                    newEinsumEquation);
+    return success();
+  }
+};
+} // namespace
+
+namespace {
+// Push up transpose from an einsum, rearranging the axes of the output tensor
+// in the einsum as needed transpose(einsum(x1, x2, ...)) -> einsum(x1, x2, ...)
+class PushUpTransposeToEinsum : public OpRewritePattern<tensorrt::TransposeOp> {
+public:
+  using OpRewritePattern<tensorrt::TransposeOp>::OpRewritePattern;
+  LogicalResult matchAndRewrite(tensorrt::TransposeOp op,
+                                PatternRewriter &rewriter) const override {
+    auto perm = op.getPermutation();
+    if (!perm.isPermutation())
+      return failure();
+
+    auto einsum = op.getInput().getDefiningOp<tensorrt::EinsumOp>();
+    if (!einsum)
+      return failure();
+
+    SmallVector<OpOperand> einsumUsers(einsum->getUsers());
+    if (einsumUsers.size() != 1)
+      return failure();
+
+    std::string einsumEquation{einsum.getEquation()};
+    size_t pos = einsumEquation.find("->");
+    if (pos == std::string::npos)
+      return failure();
+    std::string einsumEquationLhs = einsumEquation.substr(0, pos);
+    std::string einsumEquationRhs = einsumEquation.substr(pos + 2);
+
+    SmallVector<int64_t> einsumRhs;
+    for (char c : einsumEquationRhs) {
+      einsumRhs.push_back(c);
+    }
+    einsumRhs = perm.compose(einsumRhs);
+    einsumEquationRhs = "";
+    for (int i = 0; i < einsumRhs.size(); i++) {
+      einsumEquationRhs += (char)einsumRhs[i];
+    }
+
+    std::string newEinsumEquation =
+        einsumEquationLhs + "->" + einsumEquationRhs;
+
+    auto newEinsum = rewriter.create<tensorrt::EinsumOp>(
+        op.getLoc(), op.getType(), einsum.getInputs(), newEinsumEquation);
+    rewriter.replaceOp(op, newEinsum.getResult());
+    rewriter.eraseOp(einsum);
```

Comment:
Don't erase operations in patterns unless absolutely necessary (e.g. they are side-effecting). Erasing pure ops is done automatically and is error prone.